zebrad/src/commands/start/sync.rs

```diff
@@ -35,86 +36,209 @@ where
 {
     pub async fn run(&mut self) -> Result<(), Report> {
         loop {
-            if self.tip_requests.is_empty() {
+            if self.prospective_tips.is_empty() {
                 info!("populating prospective tips list");
+                // TODO(jlusby): get the block_locator from the state
                 self.obtain_tips(vec![super::GENESIS]).await?;
             }
 
             info!("extending prospective tips");
-            self.extend_tips().await?;
+            // ObtainTips Step 6
+            //
+            // If there are any prospective tips, call ExtendTips. Continue this step until there are no more prospective tips.
+            while !self.prospective_tips.is_empty() {
+                self.extend_tips().await?;
+            }
+
+            // TODO(jlusby): move this to a background task and check it for errors after each step.
             self.process_blocks().await?;
+
+            delay_for(Duration::from_secs(15)).await;
         }
     }
+
     /// Given a block_locator list fan out request for subsequent hashes to
     /// multiple peers
-    pub async fn obtain_tips(&mut self, block_locator: Vec<BlockHeaderHash>) -> Result<(), Report> {
+    //
+    // ObtainTips Step 1
+    //
+    // Query the current state to construct the sequence of hashes: handled by
+    // the caller
```

Comment:
it's better to do this inside `obtain_tips` because that logic has to be done every time `obtain_tips` is called and at no other time.